### PR TITLE
gRPC client Prometheus metrics

### LINKED ISSUE #358: gRPC client Prometheus metrics
## Is your feature request related to a problem? Please describe

The patron gRPC client comes with unary tracing, but without Prometheus metrics. Having these metrics in place could prove particularly useful in some scenarios, e.g. when the RPC network is unreliable.

## Is this a bug? Please provide steps to reproduce, a failing test etc.

Not a bug.

## Describe the solution

Add Prometheus metrics to Patron gRPC client, at least for unary. The stream implementation could be tackled separately, as we don't seem to be intercepting or tracing stream RPC. Prometheus Histogram Vector ought to be enough for now, labels should include gRPC type, target, method, and possibly status code.

## Additional context

Although most (all?) Patron clients include tracing, but most of them don't seem to be recording Prometheus metrics. introducing metrics in other clients as well, such as for AWS SNS and even SQL, a possibility which could be investigated separately.

### PR DESCRIPTION
**Pull request by christgf**
## Which problem is this PR solving?

This PR closes #358 [gRPC client Prometheus metrics](https://github.com/beatlabs/patron/issues/358)

The Patron gRPC client comes with tracing, but without Prometheus metrics. Having these metrics in place could prove particularly useful in some scenarios, e.g. when the RPC network is unreliable.

The Patron HTTP client already has Prometheus metrics - See [PR 354](https://github.com/beatlabs/patron/pull/354)

## Short description of the changes

Added Prometheus metrics to Patron gRPC client.

- Prometheus Histogram Vector with appropriate labels
- Updating Histogram accordingly for Unary RPC using existing unary interceptor
- Minor refactoring (imports, comments)
- Associated tests

## Note
The gRPC client provides a unary interceptor, which wraps RPC calls with the appropriate tracing mechanism, but _does not_ provide a stream interceptor. This PR provides Prometheus metrics for unary RPC - stream RPC would need to be handled separately.

### CODE REVIEW

File: client/grpc/grpc.go [hunk 1]
```diff
@@ -42,26 +67,36 @@ func (c *headersCarrier) Set(key, val string) {
...
+func unaryInterceptor(target string) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		span, ctx := trace.ChildSpan(ctx,
+			trace.ComponentOpName(componentName, method),
+			componentName,
+			ext.SpanKindProducer,
+		)
+		carrier := headersCarrier{Ctx: ctx}
+		err := span.Tracer().Inject(span.Context(), opentracing.TextMap, &carrier)
+		if err != nil {
```
Comment (christgf) [inline client/grpc/grpc.go:78]: @mantzas Do we really want to return here, if tracing headers cannot be injected, without invoking the RPC?

Comment (mantzas) [reply client/grpc/grpc.go:78]: Good point. This error should obviously never happen but maybe we should just log the error and continue.

File: client/grpc/grpc.go [hunk 2]
```diff
@@ -1,34 +1,59 @@
...
-	"github.com/beatlabs/patron/correlation"
-	"github.com/beatlabs/patron/trace"
-	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
+	"github.com/prometheus/client_golang/prometheus"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
```
Comment (mantzas) [inline client/grpc/grpc.go]: Empty line.

File: client/grpc/grpc_test.go [hunk 3]
```diff
@@ -97,24 +104,81 @@ func TestDialContext(t *testing.T) {
...
-	resp, err := client.SayHello(ctx, &examples.HelloRequest{Firstname: "John", Lastname: "Doe"})
-	require.NoError(t, err)
-	assert.Equal(t, "Hello John Doe", resp.GetMessage())
-	expected := map[string]interface{}{
-		"component": "grpc-client",
-		"error":     false,
-		"span.kind": ext.SpanKindEnum("producer"),
-		"version":   "dev",
+
+	tt := []struct {
```
Comment (mantzas) [inline client/grpc/grpc_test.go]: I would suggest to use a map based table test. https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

**Review by mantzas — CHANGES_REQUESTED** [review]

File: client/grpc/grpc.go [hunk 4]
```diff
@@ -77,7 +76,7 @@ func unaryInterceptor(target string) grpc.UnaryClientInterceptor {
 		carrier := headersCarrier{Ctx: ctx}
 		err := span.Tracer().Inject(span.Context(), opentracing.TextMap, &carrier)
 		if err != nil {
-			return fmt.Errorf("failed to inject tracing headers: %w", err)
+			log.Errorf("failed to inject tracing headers: %v", err)
```
Comment (mantzas) [inline client/grpc/grpc.go]:
```suggestion
			log.FromContext(ctx).Errorf("failed to inject tracing headers: %v", err)
```

**Review by mantzas — CHANGES_REQUESTED** [review]
One last small change

**Review by mantzas — APPROVED** [review]

**Comment by mantzas** [comment]
@christgf  Nice work

**Comment by christgf** [comment]
@tpaschalis @drakos74 what do you guys think?

**Review by tpaschalis — APPROVED** [review]
I'm really sorry for the belated response. LGTM, nice work, keep at it! 😄

**Pull request merged**